lib/pbench/agent/tool_meister.py

```diff
@@ -1242,35 +1273,25 @@ def stop_tools(self, data):
                 data["directory"],
                 self._directory,
             )
-            return False
+            return 1
 
-        failures = 0
-        tool_cnt = 0
-        for name in sorted(self._tools.keys()):
-            tool_cnt += 1
-            try:
-                tool = self._running_tools[name]
-            except KeyError:
-                assert (
-                    name in self.persistent_tool_names
-                ), f"tool {name} not found in list of persistent tools"
-            else:
-                try:
-                    tool.stop()
-                except Exception:
-                    self.logger.exception(
-                        "Failed to stop tool %s running in background", name
-                    )
-                    failures += 1
+        tool_cnt = len(self._running_tools)
+        failures = self._stop_running_tools()
         failures += self._wait_for_tools()
 
         # Clean up the running tools data structure explicitly ahead of
         # potentially receiving another start tools.
-        for name in sorted(self._tools.keys()):
-            try:
-                del self._running_tools[name]
-            except KeyError:
-                pass
+        if __debug__:
```

Comment:
I was trying to find a way to use `assert` here where the assert message would not be present in production code, and that none of the calculations for the assert or the message would be "leaked" in the normal code flow.

Reply:
In more detail:

1. The use of `frozenset` seems excessive here, since your sets are all quite transient; it's just taking up more page space, making it harder to read
2. Since the `assert` *statement* disappears entirely without `__debug__` there shouldn't be any point in trying to protect the expressions under an external `if __debug__`;

I'm not sure this is really any more than

```
    assert set(self.__running_tools.keys()) < set(self.__transient_tools.keys()), \
        f"The set of running tools ({','.join(sorted(self.__running_tools.keys()))}) is not a subset of the transient tools"
```

(Sure, you've avoided pulling the `keys` list(s) a second time for the report, but this isn't something we expect to happen; and the displayed value doesn't have to be a `frozenset` or even a `set`. I suspect there'll be enough information in the log that showing the values in the `assert` message probably isn't even necessary, though it's certainly convenient.)

Then again, I was just commenting. I approved, so take it or leave it. 🍤

Reply:
Let's continue any further refactoring and test enhancements in PR #2725.